src/workerd/io/io-context.c++

```diff
@@ -583,6 +585,20 @@ class IoContext::PendingEvent: public kj::Refcounted {
 };
 
 IoContext::~IoContext() noexcept(false) {
+  // Dispatch unload event on ExecutionContext if it was attached and still alive.
+  // Just like for DurableObjectState, there is no strong reference and therefore no
+  // guarantee that finalizers will run - if the GC runs first, unload will not be
+  // triggered by design.
+  if (hasUnloadHandlers) {
+    KJ_IF_SOME(weakCtx, executionContext) {
+      worker->addUnloadTask([weakCtx = weakCtx.addRef()](jsg::Lock& js) mutable {
```

Comment:
Why a weak ref? Since the IoContext is being destroyed, and the ExecutionContext is tied closely to the IoContext, it seems like it'd be pretty common for the ExecutionContext to be destroyed around this time too, making this callback unreliable. But I think you could just attach a strong ref here instead?

Reply:
Another problem here then is that in the case of durable objects, the durableobjectstate event target cannot be moved into a `jsg::Ref` during shutdown, which would be simpler with a weak ref model.

Reply:
So I support `IoContext` should have a method like `setUnloadListener(jsg::Ref<EventTarget>)`.

It's called the first time an unload event listener is added.

Reply:
The problem is that durableobjectstate unload is for the actor shutdown, so they aren't the same listener design. We are sort of implementing two separate features here.

Reply:
I've moved to using the outer waitUntilEvents on the outer io context, but am now not getting unload events in the test.

Would be great to work this through to more clearly figure out the right model here. For now I believe I've covered the feedback in https://github.com/cloudflare/workerd/pull/5211/commits/8a591dfa6c8f4c0fb97890e121288a93805aaf9c, although the implementation isn't passing the tests now.